### ISSUE TITLE
EmitterProcessor not delivering after subscriber exception

### ISSUE THREAD

**Issue by lightoze** [reactor/reactor-core#613]
Test case:
```java
public static void main(String[] args) {
    EmitterProcessor<String> proc = EmitterProcessor.create(false);
    FluxSink<String> sink = proc.sink(FluxSink.OverflowStrategy.DROP);

    Disposable bad = proc.subscribe(s -> {
        System.err.println("Bad: " + s);
        throw new IllegalStateException();
    });
    Disposable good = proc.subscribe(s -> System.err.println("Good: " + s));

    try {
        sink.next("1");
    } catch (Exception e) {
        // ignore
    }
    try {
        sink.next("2");
    } catch (Exception e) {
        // ignore
    }
    try {
        sink.next("3");
    } catch (Exception e) {
        // ignore
    }

    System.err.println(bad.isDisposed());
    System.err.println(good.isDisposed());
    System.err.println(proc.isTerminated());
    System.err.println(proc.isCancelled());
}
```
Output:
```
Bad: 1
true
false
false
false
```

The problem lies in SerializedSink - `wip` field is not decremented correctly.

**Comment by smaldini** [comment]
This needs investigating. The failing subscriber is supposed to cancel however in this case there is no error callback so a fatal error will bubble.

**Comment by lightoze** [comment]
Having an error callback helps in this case, but:
1) Seems strange to me that one subscriber failure affects others. Also a queue inside SerializedSink will grow forever after this happens.
2) For non-lambda subscribers there's no guarantee that `onNext` will never throw.

**Comment by simonbasle** [comment]
@lightoze `onError` callback could almost be considered mandatory. The `onNext` exception goes to `onError`, but if it is not implemented it has to go somewhere, so it is bubbled up and crashes the whole chain.

Can you elaborate on the queue growing forever?

@smaldini the failing subscriber does cancel, `bad.isDisposed()` returns true. Adding an error handler, `good` receives the whole sequence. I don't think this is a bug. Do you think otherwise (and if so, any way we can improve)?

**Comment by lightoze** [comment]
@simonbasle the problem is with how `WIP` counter is used in SerializedSink. If `sink.next(t)` throws, `WIP.decrementAndGet(this)` is never executed and all subsequent executions will only enqueue the object and return.

```java
sink.next(t);
if (WIP.decrementAndGet(this) == 0) {
    return;
}
``` 
Should be replaced with:
```java
try {
    sink.next(t);
} finally {
    if (WIP.decrementAndGet(this) == 0) {
        return;
    }
}
``` 

Similar problem can happen within drain loop - it is assumed that `WIP` will be always zero after `drainLoop()` completes, but it is not guaranteed in case any of callbacks throws an exception.

I also believe it is not the only place where similar approach is used and the code assumes that callbacks will never throw an exception.

**Comment by smaldini** [comment]
Per reactivestream specification onNext should never throw in effect as we protect the null input and if it throw, it is a fatal exception with fatal consequences (see Exceptions.throwIfFatal). Now we can look at input and output of a flow to mitigate that, for instance in the sink or in the lambda subscriber at the end directly.

**Comment by smaldini** [comment]
fixed by #876 in serialized sink

**Comment by smaldini** [comment]
Note that the rationale to fix it only here for now is that serialized sink is our safest emitter component, the rest must obey reactive streams spec (processor or push sink from Flux.push), and  consuming with reactor subscribers like BaseSubscriber or lambda-based subscribe protects against this via cancel. The test had to be a wrongly written Subscriber to fail.
